### PR TITLE
[Gmail Single User] Metrics

### PR DESCRIPTION
**Pull request by shmuel44**
## Related Issues
fixes: [link](https://jira-dc.paloaltonetworks.com/browse/CIAC-10620) to the issue

## Description
Dashboards 
- Added a widget showing API execution metrics for the **Gmail Single User** integration in the `API Execution Metric` dashboard.

Gmail single user
 - Added support for API Execution Metric reporting.
 - Added a new widget for the `API Execution Metrics` dashboard
 - Removed the `BETA` mention from the integration image.
 
 
image

### CODE REVIEW

File: Packs/CommonDashboards/ReleaseNotes/1_4_4.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+
+#### Dashboards
+
+##### API Execution Metrics
+
+Added a widget to provide API execution metrics for the **Gmail Single User** integration.
```
Comment (dorschw) [inline Packs/CommonDashboards/ReleaseNotes/1_4_4.md:6]:
```suggestion
Added a widget showing API execution metrics for the **Gmail Single User** integration.
```

File: Packs/CommonDashboards/pack_metadata.json [hunk 2]
```diff
@@ -2,7 +2,7 @@
     "name": "Common Dashboards",
     "description": "Frequently used dashboards pack.",
     "support": "xsoar",
-    "currentVersion": "1.4.3",
+    "currentVersion": "1.4.4",
```
Comment (dorschw) [inline Packs/CommonDashboards/pack_metadata.json:5]:
```suggestion
    "currentVersion": "1.5.0",
```

File: Packs/GmailSingleUser/Integrations/GmailSingleUser/GmailSingleUser.py [hunk 3]
```diff
@@ -49,11 +50,67 @@
     'Accept': 'application/json',
 }
 
-
+EXECUTION_METRICS = ExecutionMetrics()
 ''' HELPER FUNCTIONS '''
 
 
+def execute_gmail_action(service, action: str, action_kwargs: dict) -> dict:
```
Comment (dorschw) [inline Packs/GmailSingleUser/Integrations/GmailSingleUser/GmailSingleUser.py:57]: what's the type of `service`?

File: Packs/GmailSingleUser/Integrations/GmailSingleUser/GmailSingleUser.py [hunk 4]
```diff
@@ -49,11 +50,67 @@
...
+        HttpError: If an error occurs from the Gmail API request.
+        Exception: If a general error occurs during the function execution.
+
+    Note:
+        This function updates execution metrics counters based on the outcome of the API call.
+    """
+    try:
+        match action:
+            case "send":
+                return service.users().messages().send(**action_kwargs).execute()
```
Comment (dorschw) [inline Packs/GmailSingleUser/Integrations/GmailSingleUser/GmailSingleUser.py:82]: why is this one `return` and the others save a `result?` does it have no (interesting) return value?

File: Packs/GmailSingleUser/Integrations/GmailSingleUser/GmailSingleUser.py [hunk 5]
```diff
@@ -49,11 +50,67 @@
...
+        if error.status_code == 429:
+            EXECUTION_METRICS.quota_error += 1
+        elif error.reason == 'Unauthorized':
+            EXECUTION_METRICS.auth_error += 1
+        else:
+            EXECUTION_METRICS.general_error += 1
+        raise error
+    except Exception as error:
+        EXECUTION_METRICS.general_error += 1
+        raise error
```
Comment (dorschw) [inline Packs/GmailSingleUser/Integrations/GmailSingleUser/GmailSingleUser.py:102]:
```suggestion
    except Exception:
        EXECUTION_METRICS.general_error += 1
        raise
```

File: Packs/GmailSingleUser/Integrations/GmailSingleUser/GmailSingleUser.py [hunk 6]
```diff
@@ -1101,22 +1159,30 @@ def fetch_incidents(client: Client):
     max_results = MAX_FETCH
     if MAX_FETCH > 200:
         max_results = 200
-    LOG(f'GMAIL: fetch parameters: user: {user_key} query={query}'
-        f' fetch time: {last_fetch} page_token: {page_token} max results: {max_results}')
-    result = service.users().messages().list(
-        userId=user_key, maxResults=max_results, pageToken=page_token, q=query).execute()
+    demisto.debug(f'GMAIL: fetch parameters: user: {user_key} query={query}'
```
Comment (dorschw) [inline Packs/GmailSingleUser/Integrations/GmailSingleUser/GmailSingleUser.py:1162]:
```suggestion
    demisto.debug(f'GMAIL: fetch parameters: user: {user_key} {query=}'
```

File: Packs/GmailSingleUser/Integrations/GmailSingleUser/GmailSingleUser.py [hunk 7]
```diff
@@ -1133,7 +1199,7 @@ def fetch_incidents(client: Client):
             demisto.info(
                 f'skipped incident with lower date: {occurred} than fetch: {last_fetch} name: {incident.get("name")}')
 
-    demisto.info('extract {} incidents'.format(len(incidents)))
+    demisto.info(f'extract {len(incidents)} incidents')
```
Comment (dorschw) [inline Packs/GmailSingleUser/Integrations/GmailSingleUser/GmailSingleUser.py:1202]:
```suggestion
    demisto.info(f'extracted {len(incidents)} incidents')
```

File: Packs/GmailSingleUser/Integrations/GmailSingleUser/GmailSingleUser.py [hunk 8]
```diff
@@ -1209,9 +1275,10 @@ def main():  # pragma: no cover
             sys.exit(0)
         if command in commands:
             demisto.results(commands[command](client))
-        # Log exceptions
     except Exception as e:
         return_error(f'An error occurred: {e}', error=e)
+    finally:
+        return_metrics()
```
Comment (dorschw) [inline Packs/GmailSingleUser/Integrations/GmailSingleUser/GmailSingleUser.py:1281]: nice!

File: Packs/GmailSingleUser/Integrations/GmailSingleUser/GmailSingleUser.py [hunk 9]
```diff
@@ -49,11 +50,67 @@
...
+    except Exception as error:
+        EXECUTION_METRICS.general_error += 1
+        raise error
+    EXECUTION_METRICS.success += 1
+    return result
+
+
+def return_metrics():
+    if EXECUTION_METRICS.metrics is not None and ExecutionMetrics.is_supported():
+        return_results(EXECUTION_METRICS.metrics)
```
Comment (dorschw) [inline Packs/GmailSingleUser/Integrations/GmailSingleUser/GmailSingleUser.py:109]:
```suggestion
    if EXECUTION_METRICS.metrics is not None and ExecutionMetrics.is_supported():
        return_results(EXECUTION_METRICS.metrics)
    else:
        demisto.debug("Not returning metrics. Either metrics are not supported in this XSOAR version, or none were collected")
```

**Review by dorschw — COMMENTED** [review]
Very nice. See the comments.

File: Packs/GmailSingleUser/Integrations/GmailSingleUser/GmailSingleUser.py (hunk 3 above)
Comment (shmuel44) [reply Packs/GmailSingleUser/Integrations/GmailSingleUser/GmailSingleUser.py:57]: we don't have a type hint for this...
Maybe we need to add stub lib for this
https://github.com/henribru/google-api-python-client-stubs 🤔

File: Packs/GmailSingleUser/Integrations/GmailSingleUser/GmailSingleUser.py (hunk 4 above)
Comment (shmuel44) [reply Packs/GmailSingleUser/Integrations/GmailSingleUser/GmailSingleUser.py:82]: My mistake i will fix it...

File: Packs/GmailSingleUser/Integrations/GmailSingleUser/GmailSingleUser.py (hunk 8 above)
Comment (shmuel44) [reply Packs/GmailSingleUser/Integrations/GmailSingleUser/GmailSingleUser.py:1281]: 😊

File: Packs/GmailSingleUser/Integrations/GmailSingleUser/GmailSingleUser.py (hunk 3 above)
Comment (dorschw) [reply Packs/GmailSingleUser/Integrations/GmailSingleUser/GmailSingleUser.py:57]: True, but it's pretty much blocked by [CIAC-10534](https://jira-dc.paloaltonetworks.com/browse/CIAC-10534) so let's let it be for now.

**Review by dorschw — COMMENTED** [review]
Very nice

File: Packs/CommonDashboards/Dashboards/dashboard-APIExecutionMetrics.json [hunk 10]
```diff
@@ -1332,6 +1354,130 @@
...
+                "vcShouldIgnore": false,
+                "vcShouldKeepItemLegacyProdMachine": false,
+                "commitMessage": "",
+                "shouldCommit": false,
+                "Cache": null,
+                "name": "API Call Results For Gmail Single User",
+                "prevName": "API Call Results For Gmail Single User",
+                "dataType": "metrics",
+                "widgetType": "line",
+                "query": "type:integration  and name:\"Gmail Single User\"",
```
Comment (shmuel44) [inline Packs/CommonDashboards/Dashboards/dashboard-APIExecutionMetrics.json:1396]:
```suggestion
                "query": "type:integration and name:\"Gmail Single User\"",
```

File: Packs/CommonDashboards/Dashboards/dashboard-APIExecutionMetrics.json [hunk 11]
```diff
@@ -1340,4 +1486,4 @@
     "fromVersion": "6.8.0",
     "description": "",
     "isPredefined": true
-}
+}
```
Comment (shmuel44) [inline Packs/CommonDashboards/Dashboards/dashboard-APIExecutionMetrics.json:1489]:
```suggestion
}

```

File: Packs/GmailSingleUser/Widgets/Widget-API_Call_Results_for_Gmail_Single_User.json [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+        ],
+        "keys": [
+            "sum|totalAPICalls"
+        ],
+        "timeFrame": "hours",
+        "valuesFormat": "abbreviated",
+        "xAxisLabel": "Time",
+        "yAxisLabel": "Request Counts"
+    },
+    "query": "type:integration  and name:\"Gmail Single User\"",
```
Comment (shmuel44) [inline Packs/GmailSingleUser/Widgets/Widget-API_Call_Results_for_Gmail_Single_User.json:73]:
```suggestion
    "query": "type:integration and name:\"Gmail Single User\"",
```

**Review by dorschw — APPROVED** [review]

**Comment by ShirleyDenkberg** [comment]
@dorschw @altmannyarden @idovandijk Doc review completed.

**Pull request merged**